# SQL project

###### Project plan
<p>

   - General information
       - Project description
       - Goals of the study
       - Imports and installations
       - Connecting DB
       - Defining functions
       - Study the tables
       - Checking missings
       - Checking duplicates

   - EDA
       - Number of books released after January 1, 2000
       - Number of user reviews and the average rating for each book
       - Publisher that has released the greatest number of books with 50+ pages
       - the author with the highest average book rating
       - average number of text reviews among users who rated 50+ books

   - Summary
<p>

### General information

###### Project description

Database of one of the services competing in the market of reading apps containing data on books, publishers, authors, and customer ratings and reviews of books, will be employed to generate a value proposition for a new product.

###### Goals of the study

Tasks are:

    Find the number of books released after January 1, 2000.
    Find the number of user reviews and the average rating for each book.
    Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).
    Identify the author with the highest average book rating (look only at books with at least 50 ratings).
    Find the average number of text reviews among users who rated more than 50 books.

###### Imports and installations

In [1]:
pip install nb_black

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [2]:
pip install sql-formatter

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [3]:
%load_ext nb_black

<IPython.core.display.Javascript object>

In [4]:
import pandas as pd
import sqlalchemy
from sqlalchemy import create_engine
from sql_formatter.core import format_sql

<IPython.core.display.Javascript object>

###### Connecting DB, defining engine

In [5]:
db_config = {
    "user": "praktikum_student",
    "pwd": "Sdf4$2;d-d30pp",
    "host": "rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net",
    "port": 6432,
    "db": "data-analyst-final-project-db",
}
connection_string = "postgresql://{}:{}@{}:{}/{}".format(
    db_config["user"],
    db_config["pwd"],
    db_config["host"],
    db_config["port"],
    db_config["db"],
)
engine = create_engine(connection_string, connect_args={"sslmode": "require"})

<IPython.core.display.Javascript object>

Thank you. I rly like Python, but indentation syntax can be annoying.
<br>Probably autoformatting is a good idea. So let me try autoformat nb_black tool here.
<p>
<br>nb_black worked easily, and it took less than a minute to install and apply, it automatically fixed every Python line, but it mishandles the SQL part. 
   <p> 
    <br>Autoformatting SQL inside the Python code was truly a challenge. And I was really close today: found Jupyter magic code to replace notebook cells content, made it access SQL code with a simple regex, but after numerous attempts failed to make it work with format_sql of the sql_formatter lib, so I ended up with manual use of the format_sql function this time. It is definitely achievable but pretty tricky, will try again soon. Afterall, it is still not 100% PEP8 compliant. Anyways, it was not the project task, did it our of curiosity.

###### Defining functions

In [6]:
def display_query(query, x, full: bool):
    df = pd.io.sql.read_sql(query, con=engine)
    if full == True:
        display(
            df.head(x), df.sample(x), df.tail(x), df.info(), df.describe(include="all")
        )
    else:
        display(df.head(x), df.sample(x), df.tail(x))


setattr(pd.DataFrame, "display_query", display_query)

<IPython.core.display.Javascript object>

Super simple but pretty useful. I often use .head(), .sample() and .tail() together to observe data. I should have come up with this solution earlier to use in the projects where I called these functions frequently. This particular function takes SQL query, number of rows for tail, sample & head, and boolean showing .info() and .describe(include='all') if True

###### Study the tables

In [7]:
display_query(
    """
SELECT * 
FROM books
""",
    5,
    True,
)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
book_id             1000 non-null int64
author_id           1000 non-null int64
title               1000 non-null object
num_pages           1000 non-null int64
publication_date    1000 non-null object
publisher_id        1000 non-null int64
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

book_id  author_id                                              title  \
259      260        231               Foundation and Earth (Foundation #5)   
762      763        211                                  The Invisible Man   
534      535        522             Ranma ½  Vol. 1 (Ranma ½ (US 2nd)  #1)   
455      456          3  Murder on the Orient Express (Hercule Poirot  ...   
148      149        189                      Chronicle of a Death Foretold   

     num_pages publication_date  publisher_id  
259        500       2004-08-31           279  
762        192       2002-09-03           267  
534        304       2003-05-07           303  
455        322       2004-08-31            45  
148        120       2003-10-07           309

book_id  author_id                                              title  \
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143

None

book_id    author_id                title   num_pages  \
count   1000.000000  1000.000000                 1000  1000.00000   
unique          NaN          NaN                  999         NaN   
top             NaN          NaN  Memoirs of a Geisha         NaN   
freq            NaN          NaN                    2         NaN   
mean     500.500000   320.417000                  NaN   389.11100   
std      288.819436   181.620172                  NaN   229.39014   
min        1.000000     1.000000                  NaN    14.00000   
25%      250.750000   162.750000                  NaN   249.00000   
50%      500.500000   316.500000                  NaN   352.00000   
75%      750.250000   481.000000                  NaN   453.00000   
max     1000.000000   636.000000                  NaN  2690.00000   

       publication_date  publisher_id  
count              1000   1000.000000  
unique              618           NaN  
top          2004-06-01           NaN  
freq                 10           NaN  
mean                NaN    171.270000  
std                 NaN     99.082685  
min                 NaN      1.000000  
25%                 NaN     83.000000  
50%                 NaN    177.500000  
75%                 NaN    258.000000  
max                 NaN    340.000000

<IPython.core.display.Javascript object>

OK. 1000 entries. 

In [8]:
display_query(
    """
SELECT * 
FROM authors
""",
    5,
    False,
)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

author_id                                             author
173        174  Faye Perozich/Anne Rice/John Bolton/Daerick Gröss
547        548                           Stephen King/Ned Dameron
368        369                           Lisa Gardner/Anna Fields
443        444  Niccolò Machiavelli/Rufus Goodwin/Benjamin Mar...
620        621                                    William Golding

author_id                                author
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

<IPython.core.display.Javascript object>

636 entries here

In [9]:
display_query(
    """
SELECT * 
FROM publishers
""",
    5,
    False,
)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

publisher_id                                   publisher
220           221  Perennial / William Morrow / HarperCollins
62             63                                 City Lights
246           247                             Riverhead Books
186           187                                     NAL Jam
205           206                                    Pantheon

publisher_id                   publisher
335           336  Workman Publishing Company
336           337                  Wyatt Book
337           338       Yale University Press
338           339                    Yearling
339           340              Yearling Books

<IPython.core.display.Javascript object>

340 entries

In [10]:
display_query(
    """
SELECT * 
FROM ratings
""",
    5,
    True,
)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
rating_id    6456 non-null int64
book_id      6456 non-null int64
username     6456 non-null object
rating       6456 non-null int64
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

rating_id  book_id   username  rating
126         127       28     paul88       3
524         525       79  richard89       5
3758       3759      622   urussell       3
1916       1917      302  bobbyhood       5
5596       5597      863    npowers       4

rating_id  book_id        username  rating
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

None

rating_id      book_id    username       rating
count   6456.000000  6456.000000        6456  6456.000000
unique          NaN          NaN         160          NaN
top             NaN          NaN  martinadam          NaN
freq            NaN          NaN          56          NaN
mean    3228.500000   510.574195         NaN     3.928284
std     1863.831001   284.141636         NaN     0.943303
min        1.000000     1.000000         NaN     1.000000
25%     1614.750000   291.000000         NaN     3.000000
50%     3228.500000   506.000000         NaN     4.000000
75%     4842.250000   750.000000         NaN     5.000000
max     6456.000000  1000.000000         NaN     5.000000

<IPython.core.display.Javascript object>

6456 ratings from 1 to 5

In [11]:
display_query(
    """
SELECT * 
FROM reviews
""",
    5,
    False,
)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

review_id  book_id         username  \
2032       2031      733  patriciajohnson   
1146       1146      420      mariokeller   
1758       1757      636          fconrad   
2538       2539      907           xdavis   
84           84       30          rbender   

                                                   text  
2032  Yes something factor should father quickly can...  
1146  Live reason street treat western kitchen item....  
1758  Dinner south level expert someone. American ev...  
2538  Serious hospital after ask maybe. Stop learn a...  
84    Adult always somebody oil together room. Speci...

review_id  book_id        username  \
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...

<IPython.core.display.Javascript object>

2792 reviews. <br>Some data types could be optimized, but I don't feel like it's too relevant here

###### Checking missings 

In [12]:
def null_finder(df):
    nulls_prc = pd.DataFrame((df.isnull().sum()) * 100 / df.shape[0]).reset_index()
    nulls_prc.columns = ["Column Name", "Null Values Percent"]
    nulls_num = pd.DataFrame(df.isnull().sum()).reset_index()
    nulls_num.columns = ["Column Name", "Null Values"]
    nulls = pd.merge(nulls_num, nulls_prc, on="Column Name")
    return nulls


for query in ["books", "authors", "publishers", "ratings", "reviews"]:
    display(null_finder(pd.io.sql.read_sql(query, con=engine)))

Column Name  Null Values  Null Values Percent
0           book_id            0                  0.0
1         author_id            0                  0.0
2             title            0                  0.0
3         num_pages            0                  0.0
4  publication_date            0                  0.0
5      publisher_id            0                  0.0

Column Name  Null Values  Null Values Percent
0   author_id            0                  0.0
1      author            0                  0.0

Column Name  Null Values  Null Values Percent
0  publisher_id            0                  0.0
1     publisher            0                  0.0

Column Name  Null Values  Null Values Percent
0   rating_id            0                  0.0
1     book_id            0                  0.0
2    username            0                  0.0
3      rating            0                  0.0

Column Name  Null Values  Null Values Percent
0   review_id            0                  0.0
1     book_id            0                  0.0
2    username            0                  0.0
3        text            0                  0.0

<IPython.core.display.Javascript object>

No missings detected 

###### Checking duplicates

In [13]:
for query in ["books", "authors", "publishers", "ratings", "reviews"]:
    display(pd.io.sql.read_sql(query, con=engine).duplicated().any())

False

False

False

False

False

<IPython.core.display.Javascript object>

No duplicated rows identified

### 3. EDA

###### Number of books released after January 1, 2000

In [14]:
print(
    (
        pd.io.sql.read_sql(
            """
SELECT count(*)
FROM   books
WHERE  publication_date > '2000-01-01'
                          """,
            con=engine,
        )
    ).at[0, "count"],
    "books released after January 1st 2000",
)

819 books released after January 1st 2000


<IPython.core.display.Javascript object>

###### Number of user reviews and the average rating for each book

In [15]:
display_query(
    """
SELECT title,
       avg(ratings.rating) as mean_rating,
       count(distinct review_id) as total_reviews
FROM   books
    INNER JOIN ratings
        ON ratings.book_id = books.book_id
    INNER JOIN reviews
        ON reviews.book_id = books.book_id
GROUP BY books.book_id
ORDER BY mean_rating
""",
    5,
    True,
)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 994 entries, 0 to 993
Data columns (total 3 columns):
title            994 non-null object
mean_rating      994 non-null float64
total_reviews    994 non-null int64
dtypes: float64(1), int64(1), object(1)
memory usage: 23.4+ KB


title  mean_rating  \
0                               Harvesting the Heart         1.50   
1                  His Excellency: George Washington         2.00   
2                                              Junky         2.00   
3                                      Drowning Ruth         2.00   
4  The World Is Flat: A Brief History of the Twen...         2.25   

   total_reviews  
0              2  
1              2  
2              2  
3              3  
4              3

title  mean_rating  \
987                          Triptych (Will Trent  #1)     5.000000   
532      A Drink Before the War (Kenzie & Gennaro  #1)     4.000000   
420                                  Lord of the Flies     3.901408   
663                       Their Eyes Were Watching God     4.125000   
596  Complications: A Surgeon's Notes on an Imperfe...     4.000000   

     total_reviews  
987              2  
532              2  
420              5  
663              4  
596              2

title  mean_rating  \
989                               My Name Is Asher Lev          5.0   
990  The Demon-Haunted World: Science as a Candle i...          5.0   
991                      Emily of New Moon (Emily  #1)          5.0   
992                  Misty of Chincoteague (Misty  #1)          5.0   
993                                    Light in August          5.0   

     total_reviews  
989              2  
990              2  
991              2  
992              2  
993              2

None

title  mean_rating  total_reviews
count                   994   994.000000     994.000000
unique                  993          NaN            NaN
top     Memoirs of a Geisha          NaN            NaN
freq                      2          NaN            NaN
mean                    NaN     3.898699       2.809859
std                     NaN     0.562190       1.055873
min                     NaN     1.500000       1.000000
25%                     NaN     3.500000       2.000000
50%                     NaN     4.000000       3.000000
75%                     NaN     4.333333       3.000000
max                     NaN     5.000000       7.000000

<IPython.core.display.Javascript object>

That's it. 994 entries. rating between 1.5 and 5, number of reviews between 1 and 7.

###### Publisher that has released the greatest number of books with more than 50 pages

In [16]:
display_query(
    """
SELECT publisher,
       count(distinct books.book_id) as books
FROM   publishers
    INNER JOIN books
        ON books.publisher_id = publishers.publisher_id
WHERE  books.num_pages > 50
GROUP BY publisher
ORDER BY books
""",
    5,
    False,
)

publisher  books
0                         Deodand      1
1                        Ace Book      1
2              Everyman's Library      1
3                    Faber  Faber      1
4  Farrar  Straus and Giroux (NY)      1

publisher  books
176            Corgi Childrens      1
183           Delta Publishing      1
246     Knopf Publishing Group      3
320  Little  Brown and Company     12
328           Ballantine Books     19

publisher  books
329                    Bantam     19
330          Penguin Classics     24
331  Grand Central Publishing     25
332                   Vintage     31
333             Penguin Books     42

<IPython.core.display.Javascript object>

Penguin Books is the top publisher here, 42 books of 50+ pages

###### Identify the author with the highest average book rating (50+ rated only)

In [17]:
display_query(
    """
SELECT author,
       avg(rating) as mean_rating
FROM   authors
    INNER JOIN books
        ON books.author_id = authors.author_id
    INNER JOIN ratings
        ON ratings.book_id = books.book_id
GROUP BY author having count (rating) > 50
ORDER BY mean_rating;
""",
    5,
    False,
)

author  mean_rating
0                           John Steinbeck     3.643836
1                          Stephenie Meyer     3.662500
2  George Orwell/Boris Grabnar/Peter Škerl     3.729730
3                               Lois Lowry     3.738462
4                                Dan Brown     3.741259

author  mean_rating
6   Paulo Coelho/Alan R. Clarke/Özdemir İnce     3.789474
9                            Sophie Kinsella     3.877193
14                              Stephen King     4.009434
4                                  Dan Brown     3.741259
2    George Orwell/Boris Grabnar/Peter Škerl     3.729730

author  mean_rating
18           Roald Dahl/Quentin Blake     4.209677
19                     J.R.R. Tolkien     4.240964
20  Markus Zusak/Cao Xuân Việt Khương     4.264151
21                    Agatha Christie     4.283019
22         J.K. Rowling/Mary GrandPré     4.288462

<IPython.core.display.Javascript object>

J.K. Rowling/Mary GrandPré have the highest average rating of ~4.3

###### Find the average number of text reviews among users who rated more than 50 books

In [18]:
print(
    (
        pd.io.sql.read_sql(
            """
SELECT round(avg(rev_cnt))
FROM   (SELECT count(review_id) as rev_cnt
        FROM   reviews
        WHERE  username in (SELECT username
                            FROM   ratings
                            GROUP BY username having count(rating_id) > 50)
        GROUP BY username) as x
            """,
            con=engine,
            coerce_float=False,
        )
    ).at[0, "round"],
    "is the average number of text reviews among users who rated more than 50 books",
)

24 is the average number of text reviews among users who rated more than 50 books


<IPython.core.display.Javascript object>

### Summary

819 books released after January 1st 2000
<br>Number of user reviews and the average rating for each book shown
<br>Penguin Books is the top publisher here, 42 books of 50+ pages
<br>J.K. Rowling/Mary GrandPré have the highest average rating of ~4.3
<br>24 is the average number of text reviews among users who rated more than 50 books
<br>
<br>Made an effort to use functions as much as I could in this project as Notion hint suggested.